### PR TITLE
Part 4 of nullable and modernization

### PR DESCRIPTION
**Pull request by erikmav**
For #124

### CODE REVIEW

**Comment by erikmav** [comment]
For #124

File: src/log4net/Appender/SmtpAppender.cs [hunk 1]
```diff
@@ -375,10 +282,10 @@ protected override void SendBuffer(LoggingEvent[] events)
       // appender. This frees us from needing to synchronize again.
       try
       {
-        using StringWriter writer = new StringWriter(System.Globalization.CultureInfo.InvariantCulture);
+        using var writer = new StringWriter(System.Globalization.CultureInfo.InvariantCulture);
 
-        string t = Layout.Header;
-        if (t != null)
+        string? t = Layout?.Header;
+        if (t is not null)
```
Comment (FreeAndNil) [inline src/log4net/Appender/SmtpAppender.cs:288]:
```suggestion
        if (Layout?.Header is string header)
```

File: src/log4net/Appender/SmtpAppender.cs [hunk 2]
```diff
@@ -389,8 +296,8 @@ protected override void SendBuffer(LoggingEvent[] events)
           RenderLoggingEvent(writer, events[i]);
         }
 
-        t = Layout.Footer;
-        if (t != null)
+        t = Layout?.Footer;
+        if (t is not null)
```
Comment (FreeAndNil) [inline src/log4net/Appender/SmtpAppender.cs:300]: same here

File: src/log4net/Appender/SmtpAppender.cs [hunk 3]
```diff
@@ -547,15 +410,14 @@ public enum SmtpAuthentication
...
+    // Allow semicolon delimiters for backward compatibility.
+    private static readonly char[] ADDRESS_DELIMITERS = { ',', ';' };
 
     /// <summary>
     ///   trims leading and trailing commas or semicolons
     /// </summary>
-    private static string MaybeTrimSeparators(string s)
+    private static string? MaybeTrimSeparators(string? s)
     {
       return string.IsNullOrEmpty(s) ? s : s.Trim(ADDRESS_DELIMITERS);
```
Comment (FreeAndNil) [inline src/log4net/Appender/SmtpAppender.cs:422]:
```suggestion
      return s?.Trim(ADDRESS_DELIMITERS);
```

File: src/log4net/Core/LogImpl.cs [hunk 4]
```diff
@@ -247,7 +241,7 @@ public virtual void DebugFormat(string format, object arg0)
     {
       if (IsDebugEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0 }), null);
+        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:244]:
```suggestion
        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, [arg0]), null);
```

File: src/log4net/Core/LogImpl.cs [hunk 5]
```diff
@@ -278,7 +272,7 @@ public virtual void DebugFormat(string format, object arg0, object arg1)
     {
       if (IsDebugEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1 }), null);
+        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:275]: same

File: src/log4net/Core/LogImpl.cs [hunk 6]
```diff
@@ -310,7 +304,7 @@ public virtual void DebugFormat(string format, object arg0, object arg1, object
     {
       if (IsDebugEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1, arg2 }), null);
+        Logger.Log(ThisDeclaringType, m_levelDebug, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1, arg2 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:307]: same

File: src/log4net/Core/LogImpl.cs [hunk 7]
```diff
@@ -445,7 +439,7 @@ public virtual void InfoFormat(string format, object arg0)
     {
       if (IsInfoEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0 }), null);
+        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:442]: same

File: src/log4net/Core/LogImpl.cs [hunk 8]
```diff
@@ -476,7 +470,7 @@ public virtual void InfoFormat(string format, object arg0, object arg1)
     {
       if (IsInfoEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1 }), null);
+        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:473]: same

File: src/log4net/Core/LogImpl.cs [hunk 9]
```diff
@@ -508,7 +502,7 @@ public virtual void InfoFormat(string format, object arg0, object arg1, object a
     {
       if (IsInfoEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1, arg2 }), null);
+        Logger.Log(ThisDeclaringType, m_levelInfo, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1, arg2 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:505]: same

File: src/log4net/Core/LogImpl.cs [hunk 10]
```diff
@@ -643,7 +637,7 @@ public virtual void WarnFormat(string format, object arg0)
     {
       if (IsWarnEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0 }), null);
+        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:640]: same

File: src/log4net/Core/LogImpl.cs [hunk 11]
```diff
@@ -674,7 +668,7 @@ public virtual void WarnFormat(string format, object arg0, object arg1)
     {
       if (IsWarnEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1 }), null);
+        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:671]: same

File: src/log4net/Core/LogImpl.cs [hunk 12]
```diff
@@ -706,7 +700,7 @@ public virtual void WarnFormat(string format, object arg0, object arg1, object a
     {
       if (IsWarnEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1, arg2 }), null);
+        Logger.Log(ThisDeclaringType, m_levelWarn, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1, arg2 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:703]: same

File: src/log4net/Core/LogImpl.cs [hunk 13]
```diff
@@ -841,7 +835,7 @@ public virtual void ErrorFormat(string format, object arg0)
     {
       if (IsErrorEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0 }), null);
+        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:838]: same

File: src/log4net/Core/LogImpl.cs [hunk 14]
```diff
@@ -872,7 +866,7 @@ public virtual void ErrorFormat(string format, object arg0, object arg1)
     {
       if (IsErrorEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1 }), null);
+        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:869]: same

File: src/log4net/Core/LogImpl.cs [hunk 15]
```diff
@@ -904,7 +898,7 @@ public virtual void ErrorFormat(string format, object arg0, object arg1, object
     {
       if (IsErrorEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1, arg2 }), null);
+        Logger.Log(ThisDeclaringType, m_levelError, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1, arg2 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:901]: same

File: src/log4net/Core/LogImpl.cs [hunk 16]
```diff
@@ -1039,7 +1033,7 @@ public virtual void FatalFormat(string format, object arg0)
     {
       if (IsFatalEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0 }), null);
+        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:1036]: same

File: src/log4net/Core/LogImpl.cs [hunk 17]
```diff
@@ -1070,7 +1064,7 @@ public virtual void FatalFormat(string format, object arg0, object arg1)
     {
       if (IsFatalEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1 }), null);
+        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:1067]: same

File: src/log4net/Core/LogImpl.cs [hunk 18]
```diff
@@ -1102,7 +1096,7 @@ public virtual void FatalFormat(string format, object arg0, object arg1, object
     {
       if (IsFatalEnabled)
       {
-        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new object[] { arg0, arg1, arg2 }), null);
+        Logger.Log(ThisDeclaringType, m_levelFatal, new SystemStringFormat(CultureInfo.InvariantCulture, format, new[] { arg0, arg1, arg2 }), null);
```
Comment (FreeAndNil) [inline src/log4net/Core/LogImpl.cs:1099]: same

File: src/log4net/Core/StackFrameItem.cs [hunk 19]
```diff
@@ -24,40 +25,40 @@
...
-        m_lineNumber = frame.GetFileLineNumber().ToString(System.Globalization.NumberFormatInfo.InvariantInfo);
-        m_fileName = frame.GetFileName();
+        LineNumber = frame.GetFileLineNumber().ToString(System.Globalization.NumberFormatInfo.InvariantInfo);
+        FileName = frame.GetFileName();
+
         // get method values
-        MethodBase method = frame.GetMethod();
-        if (method != null)
+        MethodBase? method = frame.GetMethod();
+        if (method is not null)
```
Comment (FreeAndNil) [inline src/log4net/Core/StackFrameItem.cs:54]:
```suggestion
        if (frame.GetMethod() is MethodBase method)
```

File: src/log4net/Core/StackFrameItem.cs [hunk 20]
```diff
@@ -24,40 +25,40 @@
...
         {
-          if (method.DeclaringType != null)
-            m_className = method.DeclaringType.FullName;
-          m_method = new MethodItem(method);
+          if (method.DeclaringType is not null)
```
Comment (FreeAndNil) [inline src/log4net/Core/StackFrameItem.cs:56]:
```suggestion
          if (method.DeclaringType?.FullName is string className)
```

File: src/log4net/DateFormatter/AbsoluteTimeDateFormatter.cs [hunk 21]
```diff
@@ -95,78 +87,52 @@ protected virtual void FormatDateWithoutMillis(DateTime dateToFormat, StringBuil
...
-        {
-          s_lastTimeStrings.Clear();
-        }
-        else
-        {
-          timeString = (string)s_lastTimeStrings[GetType()];
-        }
-
-        if (timeString == null)
+      string timeString = s_lastTimeStrings.AddOrUpdate(GetType(),
```
Comment (FreeAndNil) [inline src/log4net/DateFormatter/AbsoluteTimeDateFormatter.cs:96]: Can you please add some tests?
The whole class seems to be uncovered.

File: src/log4net/DateFormatter/DateTimeDateFormatter.cs [hunk 22]
```diff
@@ -79,28 +60,13 @@ protected override void FormatDateWithoutMillis(DateTime dateToFormat, StringBui
...
       }
-      buffer.Append(day);
-      buffer.Append(' ');
-
-      buffer.Append(m_dateTimeFormatInfo.GetAbbreviatedMonthName(dateToFormat.Month));
-      buffer.Append(' ');
-
-      buffer.Append(dateToFormat.Year);
-      buffer.Append(' ');
+      buffer
```
Comment (FreeAndNil) [inline src/log4net/DateFormatter/DateTimeDateFormatter.cs:63]: Adding a test for this would also be great.

**Review by FreeAndNil — CHANGES_REQUESTED** [review]
Thanks a lot. Just some nits and missing tests.
I've documented the breaking changes in #124

File: src/log4net/Appender/BufferingAppenderSkeleton.cs [hunk 23]
```diff
@@ -459,59 +411,59 @@ protected override void Append(LoggingEvent loggingEvent)
...
           if (m_eventMustBeFixed)
           {
             // Derive class expects fixed events
-            loggingEvent.Fix = this.Fix;
+            loggingEvent.Fix = Fix;
           }
 
           // Not buffering events, send immediately
-          SendBuffer(new LoggingEvent[] { loggingEvent });
+          SendBuffer(new[] { loggingEvent });
```
[the remark this answers is not in the document]
Comment (erikmav) [reply src/log4net/Appender/BufferingAppenderSkeleton.cs:428]: I think downlevel .NET is interfering with [] conversions so I've been avoiding them:
3>C:\log4net2\src\log4net\Appender\BufferingAppenderSkeleton.cs(429,22,429,23): error CS1026: ) expected
3>C:\log4net2\src\log4net\Appender\BufferingAppenderSkeleton.cs(429,36,429,37): error CS1002: ; expected
3>C:\log4net2\src\log4net\Appender\BufferingAppenderSkeleton.cs(429,36,429,37): error CS1513: } expected

File: src/log4net/Core/LogImpl.cs (hunk 4 above)
Comment (erikmav) [reply src/log4net/Core/LogImpl.cs:244]: Not fixing per previous note

File: src/log4net/DateFormatter/AbsoluteTimeDateFormatter.cs (hunk 21 above)
Comment (erikmav) [reply src/log4net/DateFormatter/AbsoluteTimeDateFormatter.cs:96]: There was a class already but it was explicitly excluded from compilation. Restored and added more coverage.

File: src/log4net.Tests/DateFormatter/DateTimeDateFormatterTest.cs [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+      // Tests for prepended 0 characters for 2-digit and 3-digit portions.
+      var formatter = new DateTimeDateFormatter();
+      formatter.FormatDate(new DateTime(1970, 1, 1, 1, 1, 1).AddMilliseconds(1), writer);
+      Assert.AreEqual("01 Jan 1970 01:01:01,001", sb.ToString());
+      sb.Clear();
+
+      // Non-zero-prepend case.
+      var formatter2 = new DifferentTypeNameDateTimeDateFormatter();
+      formatter.FormatDate(new DateTime(2100, 12, 30, 11, 59, 59).AddMilliseconds(100), writer);
```
Comment (FreeAndNil) [inline src/log4net.Tests/DateFormatter/DateTimeDateFormatterTest.cs:46]:
```suggestion
      formatter2.FormatDate(new DateTime(2100, 12, 30, 11, 59, 59).AddMilliseconds(100), writer);
```
Shouldn't it be formatter2.FormatDate?

**Review by FreeAndNil — COMMENTED** [review]
Thanks Erik. Just one question regarding the test.

File: src/log4net.Tests/DateFormatter/DateTimeDateFormatterTest.cs (hunk 24 above)
Comment (erikmav) [reply src/log4net.Tests/DateFormatter/DateTimeDateFormatterTest.cs:46]: Quite right! Which means when I reran the test suite it happened to work with the internal cache. Fixed in next push.

**Review by FreeAndNil — APPROVED** [review]
Thanks a lot.

**Pull request merged**